jerry-core/ecma/builtin-objects/ecma-builtin-math.c

```diff
@@ -49,596 +84,289 @@
+    if (ecma_is_value_number (*arg))
+    {
+      arg_num = ecma_get_number_from_value (*arg);
+    }
+    else
+    {
+      ecma_value_t value = ecma_op_to_number (*arg);
 
-/**
- * The Math object's 'ceil' routine
- *
- * See also:
- *          ECMA-262 v5, 15.8.2.6
- *
- * @return ecma value
- *         Returned value must be freed with ecma_free_value.
- */
-static ecma_value_t
-ecma_builtin_math_object_ceil (ecma_value_t this_arg, /**< 'this' argument */
-                               ecma_value_t arg) /**< routine's argument */
-{
-  JERRY_UNUSED (this_arg);
-  ecma_value_t ret_value = ecma_make_simple_value (ECMA_SIMPLE_VALUE_EMPTY);
+      if (ECMA_IS_VALUE_ERROR (value))
+      {
+        return value;
+      }
 
-  ECMA_OP_TO_NUMBER_TRY_CATCH (arg_num, arg, ret_value);
+      arg_num = ecma_get_number_from_value (value);
 
-  ret_value = ecma_make_number_value (DOUBLE_TO_ECMA_NUMBER_T (ceil (arg_num)));
+      ecma_fast_free_value (value);
+    }
 
-  ECMA_OP_TO_NUMBER_FINALIZE (arg_num);
-  return ret_value;
-} /* ecma_builtin_math_object_ceil */
+    if (unlikely (ecma_number_is_nan (arg_num)))
+    {
+      result_num = arg_num;
+      break;
+    }
 
-/**
- * The Math object's 'cos' routine
- *
- * See also:
- *          ECMA-262 v5, 15.8.2.7
- *
- * @return ecma value
- *         Returned value must be freed with ecma_free_value.
- */
-static ecma_value_t
-ecma_builtin_math_object_cos (ecma_value_t this_arg, /**< 'this' argument */
-                              ecma_value_t arg) /**< routine's argument */
-{
-  JERRY_UNUSED (this_arg);
-  ecma_value_t ret_value = ecma_make_simple_value (ECMA_SIMPLE_VALUE_EMPTY);
+     /* both numbers are zeroes */
+    if (ecma_number_is_zero (arg_num)
+        && ecma_number_is_zero (result_num))
+    {
+      bool is_negative = ecma_number_is_negative (arg_num);
 
-  ECMA_OP_TO_NUMBER_TRY_CATCH (arg_num, arg, ret_value);
+      if (is_max ? !is_negative : is_negative)
+      {
+        result_num = arg_num;
+      }
+    }
+    else
+    {
+      if (is_max ? (arg_num > result_num) : (arg_num < result_num))
+      {
+        result_num = arg_num;
+      }
+    }
 
-  ret_value = ecma_make_number_value (DOUBLE_TO_ECMA_NUMBER_T (cos (arg_num)));
+    arg++;
+    args_number--;
+  }
 
-  ECMA_OP_TO_NUMBER_FINALIZE (arg_num);
-  return ret_value;
-} /* ecma_builtin_math_object_cos */
+  return ecma_make_number_value (result_num);
+} /* ecma_builtin_math_object_max_min */
 
 /**
- * The Math object's 'exp' routine
+ * The Math object's 'random' routine
  *
  * See also:
- *          ECMA-262 v5, 15.8.2.8
+ *          ECMA-262 v5, 15.8.2.14
  *
  * @return ecma value
  *         Returned value must be freed with ecma_free_value.
  */
 static ecma_value_t
-ecma_builtin_math_object_exp (ecma_value_t this_arg, /**< 'this' argument */
-                              ecma_value_t arg) /**< routine's argument */
+ecma_builtin_math_object_random () /**< 'this' argument */
```

Comment:
Should have a `void` between the parentheses and the comment removed. Or alternatively a `this` parameter added :)